icu4j/main/classes/core/src/com/ibm/icu/impl/units/ComplexUnitsConverter.java

```diff
@@ -0,0 +1,126 @@
+        //              inches)
+        //              3. then, the final result will be (6 feet and 6.74016 inches)
+        unitConverters_ = new ArrayList<>();
+        for (int i = 0, n = units_.size(); i < n; i++) {
+            if (i == 0) { // first element
+                unitConverters_.add(new UnitConverter(inputUnit, units_.get(i), conversionRates));
+            } else {
+                unitConverters_.add(new UnitConverter(units_.get(i - 1), units_.get(i), conversionRates));
+            }
+        }
+    }
+
+    /**
+     * Returns true if the specified `quantity` of the `inputUnit`, expressed in terms of the biggest
+     * unit in the MeasureUnit `outputUnit`, is greater than or equal to `limit`.
+     * <p>
+     * For example, if the input unit is `meter` and the target unit is `foot+inch`. Therefore, this
+     * function will convert the `quantity` from `meter` to `foot`, then, it will compare the value in
+     * `foot` with the `limit`.
+     */
+    public boolean greaterThanOrEqual(BigDecimal quantity, BigDecimal limit) {
+        assert !units_.isEmpty();
+
+        // NOTE: First converter converts to the biggest quantity.
+        return unitConverters_.get(0).convert(quantity).multiply(EPSILON_MULTIPLIER).compareTo(limit) >= 0;
+    }
+
+    /**
+     * Returns outputMeasures which is an array with the corresponding values.
+     * - E.g. converting meters to feet and inches.
+     * 1 meter --> 3 feet, 3.3701 inches
+     * NOTE:
+     * the smallest element is the only element that could have fractional values. And all
+     * other elements are floored to the nearest integer
+     */
+    public List<Measure> convert(BigDecimal quantity) {
+        List<Measure> result = new ArrayList<>();
+
+        for (int i = 0, n = unitConverters_.size(); i < n; ++i) {
+            quantity = (unitConverters_.get(i)).convert(quantity);
+
+            if (i < n - 1) {
+                // The double type has 15 decimal digits of precision. For choosing
+                // whether to use the current unit or the next smaller unit, we
+                // therefore nudge up the number with which the thresholding
+                // decision is made. However after the thresholding, we use the
+                // original values to ensure unbiased accuracy (to the extent of
+                // double's capabilities).
+                BigDecimal newQuantity = quantity.multiply(EPSILON_MULTIPLIER).setScale(0, RoundingMode.FLOOR);
+
+                result.add(new Measure(newQuantity, units_.get(i).build()));
+
+                // Keep the residual of the quantity.
+                //   For example: `3.6 feet`, keep only `0.6 feet`
+                quantity = quantity.subtract(newQuantity);
+            } else { // LAST ELEMENT
+                result.add(new Measure(quantity, units_.get(i).build()));
+            }
+        }
+
+        return result;
```

Comment:
FYI: this commit is already in C++ upstream:

https://github.com/icu-units/icu/pull/86/commits/602e41b080fa5d18ad9491ce299938f094a13b9c

It takes care of returning "inch-and-foot", when that was the requested output unit. I don't mind whether this is part of this PR or part of a later PR, but please remember to also take care of this? I've made the change in UnitsTest::testComplexUnitConverterSorting(), I've not tested for it in new unit test, so adding only the new tests won't catch this.

Reply:
this is should be part of the glue code PR.

Reply:
It wasn't in the glue code PR. I've filed https://github.com/icu-units/icu/issues/103 to track this.